net/dns/manager.go

```diff
@@ -454,17 +463,61 @@ func (s *dnsTCPSession) handleReads() {
 // servicing DNS requests sent down it.
 func (m *Manager) HandleTCPConn(conn net.Conn, srcAddr netaddr.IPPort) {
 	s := dnsTCPSession{
-		m:            m,
-		conn:         conn,
-		srcAddr:      srcAddr,
-		responses:    make(chan []byte),
-		readClosing:  make(chan struct{}),
+		m:           m,
+		conn:        conn,
+		srcAddr:     srcAddr,
+		responses:   make(chan []byte),
+		readClosing: make(chan struct{}),
 	}
 	s.ctx, s.closeCtx = context.WithCancel(context.Background())
 	go s.handleReads()
 	s.handleWrites()
 }
 
+// HandleTCPConn implements magicDNS over DNS-over-TLS, taking a connection and
+// servicing DNS requests sent down it.
+func (m *Manager) HandleDNSoverTLSConn(conn net.Conn, srcAddr netaddr.IPPort) {
+	privKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		m.logf("[unexpected] GenerateKey: %v", err)
+		conn.Close()
+		return
+	}
+	template := x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			Organization: []string{"Tailscale MagicDNS"},
+		},
+		NotBefore: time.Now(),
+		NotAfter:  time.Now().Add(time.Hour * 24 * 180),
```

Comment:
Oh no, this is going to be terrible for tagged nodes (key expiry disabled)

Reply:
I don't know what that has to do with anything here?

Reply:
Isn't this cert going to expire after 6 months? I assume you picked 6 months because it's the same as the key expiry period?

Reply:
It's making a new cert on each TCP connection!

This was a proof of concept to see if it made Android happy.

6 months has nothing to do with key expiry or anything. It's copy/paste from some Go cert generation code elsewhere.

The latest version caches the cert, which is bad in a new way: if your Tailscale process runs for 6 months your cert can expire. But that means you're not restarting Tailscale or your phone or applying security updates for 6 months, which seems unlikely.